protocols/mdns/src/behaviour.rs

```diff
@@ -36,29 +36,63 @@ use smallvec::SmallVec;
 use std::collections::hash_map::{Entry, HashMap};
 use std::{cmp, fmt, io, net::IpAddr, pin::Pin, task::Context, task::Poll, time::Instant};
 
-#[cfg(feature = "async-io")]
-use crate::behaviour::{socket::asio::AsyncUdpSocket, timer::asio::AsyncTimer};
+/// Trait used to provide the network interface watcher to [`GenMdns`].
+pub trait WatcherProvider {
+    /// The IfWatcher type.
+    type Watcher: Stream<Item = std::io::Result<IfEvent>> + fmt::Debug + Unpin;
+
+    /// Create a new instance of the `IfWatcher` type.
+    fn new_watcher() -> Result<Self::Watcher, std::io::Error>;
+}
 
 /// The type of a [`GenMdns`] using the `async-io` implementation.
 #[cfg(feature = "async-io")]
-pub type Mdns = GenMdns<AsyncUdpSocket, AsyncTimer>;
+pub mod async_io {
+    use super::WatcherProvider;
+    use crate::behaviour::{socket::asio::AsyncUdpSocket, timer::asio::AsyncTimer};
+    use if_watch::smol::IfWatcher;
 
-#[cfg(feature = "tokio")]
-use crate::behaviour::{socket::tokio::TokioUdpSocket, timer::tokio::TokioTimer};
+    pub type Behaviour = super::Behaviour<AsyncUdpSocket, AsyncTimer>;
+
+    impl WatcherProvider for Behaviour {
+        type Watcher = IfWatcher;
+
+        fn new_watcher() -> Result<Self::Watcher, std::io::Error> {
+            IfWatcher::new()
+        }
+    }
+}
 
 /// The type of a [`GenMdns`] using the `tokio` implementation.
 #[cfg(feature = "tokio")]
-pub type TokioMdns = GenMdns<TokioUdpSocket, TokioTimer>;
+pub mod tokio {
+    use super::WatcherProvider;
+    use crate::behaviour::{socket::tokio::TokioUdpSocket, timer::tokio::TokioTimer};
+    use if_watch::tokio::IfWatcher;
+
+    pub type Behaviour = super::Behaviour<TokioUdpSocket, TokioTimer>;
+
+    impl WatcherProvider for Behaviour {
+        type Watcher = IfWatcher;
+
+        fn new_watcher() -> Result<Self::Watcher, std::io::Error> {
+            IfWatcher::new()
+        }
+    }
+}
 
 /// A `NetworkBehaviour` for mDNS. Automatically discovers peers on the local network and adds
 /// them to the topology.
 #[derive(Debug)]
-pub struct GenMdns<S, T> {
+pub struct Behaviour<S, T>
```

Comment:
Would it make sense to combine all type parameters here to a single one, i.e. `P` and have a `Provider` trait?

```rust
trait Provider {
	type Socket;
	type Timer;
	type IfWatcher;
}
```

Reply:
indeed it does! Thanks Thomas. :+1:  
We now have more than one `Provider` trait among the protocols (`tcp` and `mdns`) but as they are namedspaced by the crates I don't think it's a problem, it can be considered akin to the `tokio` and `async_io` namespaces.